### ISSUE TITLE
Suggestion: Add RimSort to package managers

### ISSUE THREAD

**Issue by Maddremor** [oceancabbage/RimSort#35]
Consider adding RimSort to package managers as to make it easier for any users to keep it up to date. While primarily relevant to Linux systems, I would personally also suggest [Scoop](https://scoop.sh/) and/or [Chocolatey](chocolatey.org/) for Windows users.

**Comment by twstagg** [comment]
@Maddremor This warrants more discussion, and would require individuals to take responsibility of certain community-facing roles regarding this here. 

Basically, we need maintainers, and a system setup so that there are others willing to step in if that assigned maintainer is unavailable.

- For 3rd party Windows package managers, Chocolatey is what I have used before and I would not be against a release here. 
    - We might consider creating a sub-task to create an actual installer of sorts for Windows users. An update system is tbd but obviously implied that we will need this as well. Just things to further consider regarding this scope.
- For Linux platform, we could also consider creating flatpak/snap packages, as opposed to maintaining packages for individual distributions. If we were to maintain for specific distributions, we could create:
    - Ubuntu/Debian packages
    - A standalone rpm package for Fedora, RHEL, etc
    - Arch pkg / AUR pkg
- For MacOS, I am thinking perhaps a normal .dmg image with the app inside for distribution.

**Comment by oceancabbage** [comment]
I definitely need to do more digging into this. I like the idea of package managers, but we could also implement an update feature directly on the client (e.g. RimPy's `Update` button). This might be a more user-friendly method for the majority of users, but I'm not sure exactly how we could set this up and if we could set it up with only the resources available on GitHub (I hope this is the case). Keeping RimSort updated is an important concern though, so I think this issue should be a priority to at least start digging into.

**Comment by joseasoler** [comment]
Packages required for getting RimSort running in Arch without using pip:

**Official repositories:** pyside2, python-toposort, python-xmltodict
**AUR:** python-steam

**Comment by Maddremor** [comment]
I'm coming at this from the user end, but from my understanding Scoop can be set to automatically generate updates from new releases using GitHub actions on the repo hosting the list of packages. This could probably be automated to push whatever is the most recent release from this repo. (Adding the qualifiers since I could just be plain wrong)

**Comment by jtljac** [comment]
I have created a bin package on the AUR that just installs the latest release: https://aur.archlinux.org/packages/rimsort-bin
This isn't perfect as the install size is almost 500MiB but it does make for an easy install

**Comment by hideki2k02** [comment]
IMO we should stick with Flatpak for Official Releases,. Its too much of a hassle to try and port/maintain RimSort in every single Package Manager avaliable as system dependencies could break anytime. 

Flatpak is plain and simple and works anywhere.

**Comment by YAMLcase** [comment]
for windows, why not have the exe install itself on first run to %APPDATA% and run from there?  no need for an installer.  Otherwise Chocolatey is the only way that should be considered for Windows package install.

**Comment by twstagg** [comment]
https://docs.python.org/3/library/multiprocessing.html

The 'spawn' and 'forkserver' start methods cannot currently be used with “frozen” executables (i.e., binaries produced by packages like PyInstaller and cx_Freeze) on Unix. The 'fork' start method does work. (This also applies to Nuitka)

Issue 1) When it does that, particularly on Linux, there can be different system libs and stuff already in use. What Nuitka does not do is allow us to produce a fully compaptible build on Linux unfortunately

Issue 2) is that on Windows and Mac, Multiprocessing can have issues when the app is "built" and not running from Python

**Comment by N3X15** [comment]
RimPy used Nuitka to make native executables.  Might be something to try.

**Comment by LionelColaso** [comment]
> RimPy used Nuitka to make native executables. Might be something to try.

we already use nuitka 
https://github.com/RimSort/RimSort/wiki/Development-Guide
![image](https://github.com/RimSort/RimSort/assets/2488703/0b9cf574-b8d9-42e9-a332-bf8ada1f63a9)

**Comment by v-evets** [comment]
> IMO we should stick with Flatpak for Official Releases,. Its too much of a hassle to try and port/maintain RimSort in every single Package Manager avaliable as system dependencies could break anytime.
> 
> Flatpak is plain and simple and works anywhere.

Flatpak is liable to pull in several GB of "runtimes", most of which duplicate things already installed through the native package manager. For reference, installing the unofficial RimPy flatpak on my system consumes an utterly ludicrous **2.9GB** of disk space.
Please do not. If you *must* use one of these newfangled "universal" ~~bloatware cannons~~ packaging formats, please use AppImage.

IMO source tarballs, instructions, and a sane sandbox-friendly (i.e. no network access at build-time or writes outside the build directory) build system would be just fine. In that case I might volunteer to maintain an ebuild for Gentoo.

> RimPy used Nuitka to make native executables. Might be something to try.
RimPy also suffered a gratuitously large install footprint, and numerous problems caused by shipping bundled libraries that were either out of date or otherwise incompatible with system software.
RimSort already has the former, and if it ever gets a "stable" release that hangs about for more than a year or two, I expect to see the latter as well.

Compiling from source for the target distribution results in a smaller, faster application with fewer runtime compatibility issues. It's been done that way for the past 25 years for a reason.

**Comment by jtljac** [comment]
> Flatpak is liable to pull in several GB of "runtimes", most of which duplicate things already installed through the native package manager. For reference, installing the unofficial RimPy flatpak on my system consumes an utterly ludicrous 2.9GB of disk space.
Please do not. If you must use one of these newfangled "universal" bloatware cannons packaging formats, please use AppImage.

This is an expected consequence of Flatpak. I see no reason not to have an official Flatpak, which ensures most users are using exactly the tested library versions thus reducing potential issue reports for what end up being library based issues.

For those who prefer a slim build, a generic set of instructions for building such should be enough, which can then be used by volunteers to setup with their respective package managers.

**Comment by Raizanad** [comment]
For a scoop json on the stable branch, someone may want to officially submit this:
```
{
    "version": "1.0.8",
    "description": "Mod manager for RimWorld",
    "homepage": "https://github.com/RimSort/RimSort",
    "license": "GPL-3.0-or-later",
    "architecture": {
        "64bit": {
            "url": "https://github.com/RimSort/RimSort/releases/download/v1.0.8/RimSort-v1.0.8-Windows_x86_64.zip",
            "hash": "935c021bcc26f5156a8f6d69a6e95de7c9bc463e788d337887b85792202659f2"
        },
        "32bit": {
            "url": "https://github.com/RimSort/RimSort/releases/download/v1.0.8/RimSort-v1.0.8-Windows_x86_64.zip",
            "hash": "935c021bcc26f5156a8f6d69a6e95de7c9bc463e788d337887b85792202659f2"
        }
    },
    "shortcuts": [
        [
            "RimSort/RimSort.exe",
            "RimSort"
        ]
    ],
    "persist": [
        "User Files"
    ],
    "checkver": {
        "github": "https://github.com/RimSort/RimSort"
    },
    "autoupdate": {
        "url": "https://github.com/RimSort/RimSort/releases/download/v$version/RimSort-v$version-Windows_x86_64.zip"
    }
}
```

**Comment by Big-Iron-Cheems** [comment]
Package added to the `scoop-games` bucket as per [this](https://github.com/Calinou/scoop-games/pull/1197) PR.

**Comment by yodatak** [comment]
I'm creating a flatpak right now for this app because it's segfault for me in fedora and I want to use it on steam deck and bazzite. I can make a WIP PR if somebody want to help

**Comment by headquarter8302** [comment]
Hey. I was wondering if we can also add a self-extracting installer like the one 7-Zip provides, so that the app can be published to the WinGet package repository.

Since RimSort on Windows is portable and doesn't need a set install location, I believe that this solution is the easiest of the bunch. I can help with the development of this if needed

**Comment by fuddlesworth** [comment]
@yodatak did this ever happen?

**Comment by yodatak** [comment]
> [@yodatak](https://github.com/yodatak) did this ever happen?

Yes flatpak is here but is not yet merged in flathub ( and i think there is a bug in todd implementation in python)

if you want to test 
its'here

https://github.com/flathub/flathub/pull/6197

**Comment by monday-j** [comment]
Not to add further complexity, but I feel that for Linux releases, building for Ubuntu specifically is a bit strange? I'd rather a standard x86 tarball and just a note stating distro-specific dependencies, it's not like (most) Linux users are allergic to reading install documentation.

Personally I feel like packaging a tarball as well as an AppImage makes the most sense, AppImage being a fairly easy way for a standard user to jump into things that's fairly distro-independent. Flatpak has its merits too especially as it can package required libraries in-app.

Of course, figuring out sane Linux packaging is always going to be a major struggle, especially as its often so opinionated 😅

**Comment by usagirei** [comment]
For anyone interested, while the flatpak isn't ready, you can use this [script](https://gist.github.com/usagirei/e40647bdc38641e3690df8b8f0909da7) i made to download the latest RimSort release and build an AppImage.

Needs `appimagetool` acessible in path, you can get it from [here](https://github.com/AppImage/appimagetool/releases) and put it on your `~/.local/bin/`

Script will fetch the latest git tag and compare with currently built AppImage. 
If different, an AppImage will be built from the Ubuntu 24.04 artifact.
Built AppImage is stored under `~/.local/share/RimSort/bin/`
A symlink under `~/.local/bin/RimSort`, and .desktop file on `~/.local/share/applications/` will also be created if needed.

**Comment by cebarks** [comment]
I'm gonna look into building a single universal macOS binary (both intel and apple silicon in a single .app) and also into the requirements for signing so we don't need to bypass gatekeeper
